File: src/main/java/xyz/nkomarn/composter/server/WorldManager.java
package xyz.nkomarn.composter.server;

import kyta.composter.Tickable;
import kyta.composter.server.MinecraftServer;
import org.jetbrains.annotations.NotNull;
import xyz.nkomarn.composter.world.ChunkIO;
import xyz.nkomarn.composter.world.World;
import xyz.nkomarn.composter.world.generator.FlatGenerator;

import java.nio.file.Path;
import java.util.*;
import java.util.concurrent.Executor;
import java.util.concurrent.ThreadLocalRandom;

public class WorldManager implements Tickable {
    private final MinecraftServer server;
    private final Path directory;

    private final HashMap<UUID, World> worlds;
    private final Executor chunkThread;

    public WorldManager(@NotNull MinecraftServer server, @NotNull Path directory) {
        this.server = server;
        this.directory = directory;
        this.worlds = new HashMap<>();
        this.chunkThread = Runnable::run; // Executors.newVirtualThreadPerTaskExecutor();// TODO configurable - Runnable::run;
    }

    public World primaryWorld() {
        return worlds.values().stream().findFirst().orElseThrow();
    }

    public void load() {
        directory.toFile().mkdirs();

        // TODO temporarily create just 1 world
        var seed = ThreadLocalRandom.current().nextLong(Long.MIN_VALUE, Long.MAX_VALUE);
        var world = new World(server, new World.Properties(
                UUID.randomUUID(),
                seed,
                new ChunkIO(server, directory.resolve("world"), chunkThread),
                FlatGenerator.INSTANCE),
                //new FlatGenerator(),
                chunkThread
        );
        worlds.put(UUID.randomUUID(), world);
    }

    public Optional<World> getWorld(@NotNull UUID uuid) {
        return Optional.ofNullable(worlds.get(uuid));
    }

    public Collection<World> getWorlds() {
        return worlds.values();
    }

    @Override
    public void tick(long currentTick) {
        worlds.values().forEach(world -> world.tick(currentTick));
    }
}

File: src/main/java/xyz/nkomarn/composter/server/WorldManager.kt
package xyz.nkomarn.composter.server

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kyta.composter.Tickable
import kyta.composter.server.MinecraftServer
import kyta.composter.world.BlockPos
import kyta.composter.world.World
import kyta.composter.world.dimension.DimensionType
import xyz.nkomarn.composter.world.ChunkIO
import xyz.nkomarn.composter.world.generator.FlatGenerator
import xyz.nkomarn.composter.world.generator.NoiseGenerator
import java.nio.file.Path
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ThreadLocalRandom
import java.util.function.Consumer

class WorldManager(
    private val server: MinecraftServer,
    private val directory: Path,
) : Tickable {
    private val worlds = mutableMapOf<DimensionType, World>()
    val primaryWorld: World
        get() = worlds[DimensionType.OVERWORLD]!!

    fun load() {
        directory.toFile().mkdirs()

        // TODO temporarily create just 1 world
        val seed = ThreadLocalRandom.current().nextLong()
        val world = World(
            server,
            World.Properties(
                DimensionType.OVERWORLD,
                seed,
                ChunkIO(server, directory.resolve("world")),
                NoiseGenerator(seed.toInt()),
                BlockPos(0, 62, 0),
            )
        )
        worlds[world.properties.type] = world
    }

    /*
     * tick all worlds in parallel.
     */
    override fun tick(currentTick: Long) {
        val latch = CountDownLatch(worlds.size)

        worlds.values.forEach {
            server.launch(Dispatchers.Default) {
                try {
                    it.tick(currentTick)
                } catch (x: Throwable) {
                    server.logger.error("encountered an error ticking world ${it.properties.type}", x)
                } finally {
                    latch.countDown()
                }
            }
        }

        latch.await()
    }
}
